GoogleForCreators/web-stories-wp
# Shapes: Add simple border option

## Context

Enables the border panel for non-rectangular shapes with the possibility to add a simple border that supports:

- A single border-width value only.
- Allows choosing color **but not** opacity for the border.
- Only a middle-positioned border (unlike for rectangular elements).

This is all tested and confirmed working with output, flip, rotation, focal changes, focal zoom, page zoom, RTL, layer opacity, and more.

### Todo

- [x] Fix media snapping bug (currently media has a hard time snapping into shapes - not sure what broke) Must be fixed in this PR!
- [x] Fix media edit mode when in non-rect bordered shape
- [x] Fix the weird scaling bug
- [ ] Figure out what happens when dropping a multi-bordered and/or semitransparent-bordered media element into a shape that does not support it (follow-up bug probably).
- [ ] File follow-up bug for the color picker opacity error text discrepancy mentioned below.

## User-facing changes
### Design panel
The new border width panel for non-rectangular shapes (note the single input, no lock button and no opacity for border color):
![Screen Shot 2022-01-22 at 00 02 47](https://user-images.githubusercontent.com/637548/150625464-f4ec8138-e110-4a5c-9b0d-a639133c85bd.png)

Note the following "bug" in the basic color picker, where transparency is not allowed (which is correct), but the error message assumes that this is because it's about setting page background color, which this isn't.
![Screen Shot 2022-01-22 at 00 02 57](https://user-images.githubusercontent.com/637548/150625517-f278000b-1f1c-4be6-83d7-4b3356d92bb4.png)

A follow-up bug should be filed when this is merged and the text should be customized for page background and non-rectangular shape/mask border-color respectively. Or it should be made generic enough to support both cases (e.g. _"The color on this element does not support opacity"_).

### Canvas
Here are the various elements with their borders increasing and decreasing again:

| ![border-rect](https://user-images.githubusercontent.com/637548/150626183-87f4cf10-bd71-4192-8979-dad461d04dbe.gif) | ![border-circ](https://user-images.githubusercontent.com/637548/150626187-05d379a9-23e3-4c69-9e44-910e5d1d8e0f.gif) | ![border-tri](https://user-images.githubusercontent.com/637548/150626190-6cc77091-6257-4fd3-b613-79da088bb215.gif) | ![border-heart](https://user-images.githubusercontent.com/637548/150626191-40b9f62e-9d0b-4808-9e04-963c68172b33.gif) |
|-|-|-|-|
| ![border-star](https://user-images.githubusercontent.com/637548/150626238-3a4b1a49-f0ff-4bb7-8e36-c263d496988b.gif) | ![border-pent](https://user-images.githubusercontent.com/637548/150626239-7e625d99-567f-4998-b721-66bcbb125e08.gif) | ![border-hex](https://user-images.githubusercontent.com/637548/150626240-456aba8e-f7f0-48be-afa3-9b6a6578f47f.gif) | ![border-blob1](https://user-images.githubusercontent.com/637548/150626241-17c91764-b6b6-4422-bcd0-814d8b8503b6.gif) |
| ![border-blob2](https://user-images.githubusercontent.com/637548/150626244-06d97e40-7714-4b9b-84cc-debd9f23f90a.gif) | ![border-blob3](https://user-images.githubusercontent.com/637548/150626246-1267537d-eef6-4e69-aaeb-21278467d0ca.gif) |![border-blob4](https://user-images.githubusercontent.com/637548/150626247-1b2f29f3-00aa-47fd-a5f7-6dbbb2c6eabb.gif) |![border-blob5](https://user-images.githubusercontent.com/637548/150626249-63ee538d-2b31-4474-a0fd-f6a66aecf265.gif) |

### Output
Compare canvas with output:
| Canvas | Output |
|-|-|
| ![Screen Shot 2022-01-22 at 00 08 49](https://user-images.githubusercontent.com/637548/150625579-99638576-4e40-4d21-adb2-6bacba3c78a2.png) | ![Screen Shot 2022-01-22 at 00 08 59](https://user-images.githubusercontent.com/637548/150625580-cfa249e6-8fba-49df-8b61-974626b8908f.png) |

## Testing Instructions

This PR can be tested by following these steps:

1. Add  a non-rectangular shape to the page
2. Select the shape and verify that you can edit the border width  from the border panel
3. verify that you can update the border width and color, but not the opacity
4. observe that you cannot toggle the border width to input multiple values
5. set layer opacity to 50% and observe that the whole element correctly fades down and there's no weirdness happening in the border overlap region

## Checklist


---

Fixes #6171
